rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,379 @@
+# keras-nlp Transformer Encoder API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com), Hongkun Yu (hongkuny@google.com)|
+| **Sponsor(s)** | Mark Omernick (momernick@google.com)|
+| **Updated**   | 2020-09-21                                           |
+
+
+## Objective
+
+We aim at providing a set of Keras layers to handle Transformer-Encoder BERT-style models.
+
+## Key Benefits
+
+BERT-style Transformer-Encoders are a state-of-art technique that powers many NLP tasks:
+
+- Single sentence classification task, e.g., sentiment analysis
+- Sentence pair classification task, e.g., next sentence prediction
+- Question answering task, e.g., SQuAD
+- Single sentence tagging task, e.g., named entity recognition
+
+With this proposal, Keras users will be able to handle the tasks above with a simple API. 
+
+## Design overview
+
+This proposal builds on the assumption that inputs are lookup indices, i.e., `tf.int64` sequences.
+Tokenization is not part of this proposal but will be our immediate next step.
+
+### Classification task
+
+Case where a user want to use a pretrained BERT encoder for sentiment analysis:
+
+```python
+# Considering a imbd review dataset
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_nlp
+import tensorflow_text as tftext
+
+imdb_reviews = tfds.load('imdb_reviews')
+train_ds = imdb_reviews['train'].batch(32)
+test_ds = imdb_reviews['test'].batch(32)
+
+# Tokenization with BertTokenizer
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
```

Comment:
The preprocessing function here is quite difficult to parse. 101 and 102 are unexplained constants. There are many steps. We should simplify.

Question: why BertTokenizer instead of TextVectorization?

Reply:
Or rather WordPieceVectorizer or some other name that describes in a high level what is actually happening here.

Reply:
Re Francois: yes 101 and 102, I updated to use `tokenize_with_offsets` instead, which will do the prepend and append logic.
BertTokenizer is used to tokenize words into sub-words, which is the word piece algorithm.

Re Alex: BertTokenizer = BasicTokenizer (uncasing, split, standardization) + WordPieceTokenizer. This is a mirroring of the current status of tf.text ops. I do agree once we have Tokenization RFC, we should name this to be keras_nlp.ops.WordPieceTokenizer that performs both. For now I'd like this example can actually be run successfully in colab.

Reply:
Not very crucial but there's typo in the comment: "Considering a imbd review dataset". It should be "Considering an imdb review dataset".